sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ConsumerSpEL.java

```diff
@@ -88,4 +109,34 @@ public long getRecordTimestamp(ConsumerRecord<byte[], byte[]> rawRecord) {
     }
     return timestamp;
   }
+
+  @SuppressWarnings("unchecked")
+  public long offsetForTime(Consumer<?, ?> consumer, TopicPartition topicPartition, Date time) {
+    if (hasOffsetsForTimes && time != null) {
+      Map<TopicPartition, Long> timestampsToSearch = new HashMap<>();
+      timestampsToSearch.put(topicPartition, time.getTime());
+      try {
+        Map offsetsByTimes = (Map) offsetsForTimesMethod.invoke(consumer, timestampsToSearch);
+        Object offsetAndTimestamp = Iterables.getOnlyElement(offsetsByTimes.values());
+
+        // 1. If no message has a timestamp that is greater than or equals to the target time,
+        //    a null will be returned.
+        // 2. If the message format version in a partition is before 0.10.0, i.e.
+        //    the messages do not have timestamps, null will be returned for that partition.
+        if (offsetAndTimestamp == null) {
+          Map<TopicPartition, Long> endOffsets =
+              (Map<TopicPartition, Long>) endOffsetsMethod.invoke(
+                  consumer, Collections.singleton(topicPartition));
+          return Iterables.getOnlyElement(endOffsets.values());
```

Comment:
case 1, what if target time is earlier than all messages in the topic, it will return the earliest message offset.

case 2. if messages don't have timestamp, actually we can not know that `Consumer.offsetsForTimes()` returns null because the messages do not have timestamps or because no message has a timestamp that is greater than or equals to the target time.
I think we can ignore these no timestamp messages because these messages are the user does not want. But need more messages to clearly say this situation.